engine/execution/ingestion/stop/stop_control.go

```diff
@@ -0,0 +1,659 @@
+// Its jut a small subset of storage.Headers for comments see storage.Headers
+type StopControlHeaders interface {
+	ByHeight(height uint64) (*flow.Header, error)
+}
+
+// NewStopControl creates new StopControl.
+//
+// We currently have no strong guarantee that the node version is a valid semver.
+// See build.SemverV2 for more details. That is why nil is a valid input for node version
+// without a node version, the stop control can still be used for manual stopping.
+func NewStopControl(
+	log zerolog.Logger,
+	exeState state.ReadOnlyExecutionState,
+	headers StopControlHeaders,
+	versionBeacons storage.VersionBeacons,
+	nodeVersion *semver.Version,
+	latestFinalizedBlock *flow.Header,
+	withStoppedExecution bool,
+	crashOnVersionBoundaryReached bool,
+) *StopControl {
+	// We should not miss block finalized events, and we should be able to handle them
+	// faster than they are produced anyway.
+	blockFinalizedChan := make(chan *flow.Header, 1000)
+
+	sc := &StopControl{
+		log: log.With().
+			Str("component", "stop_control").
+			Logger(),
+
+		blockFinalizedChan: blockFinalizedChan,
+
+		exeState:                      exeState,
+		headers:                       headers,
+		nodeVersion:                   nodeVersion,
+		versionBeacons:                versionBeacons,
+		stopped:                       withStoppedExecution,
+		crashOnVersionBoundaryReached: crashOnVersionBoundaryReached,
+		// the default is to never stop
+		stopBoundary: stopBoundary{
+			StopParameters: StopParameters{
+				StopBeforeHeight: NoStopHeight,
+			},
+		},
+	}
+
+	if sc.nodeVersion != nil {
+		log = log.With().
+			Stringer("node_version", sc.nodeVersion).
+			Bool("crash_on_version_boundary_reached",
+				sc.crashOnVersionBoundaryReached).
+			Logger()
+	}
+
+	log.Info().Msgf("Created")
+
+	cm := component.NewComponentManagerBuilder()
+	cm.AddWorker(sc.processEvents)
+	cm.AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+		sc.checkInitialVersionBeacon(ctx, ready, latestFinalizedBlock)
+	})
+
+	sc.cm = cm.Build()
+	sc.Component = sc.cm
+
+	// TODO: handle version beacon already indicating a stop
+	// right now the stop will happen on first BlockFinalized
+	// which is fine, but ideally we would stop right away.
+
+	return sc
+}
+
+// BlockFinalized is called when a block is finalized.
+//
+// This is a protocol event consumer. See protocol.Consumer.
+func (s *StopControl) BlockFinalized(h *flow.Header) {
+	s.blockFinalizedChan <- h
+}
+
+// processEvents is a worker that processes block finalized events.
+func (s *StopControl) processEvents(
+	ctx irrecoverable.SignalerContext,
+	ready component.ReadyFunc,
+) {
+	ready()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case h := <-s.blockFinalizedChan:
+			s.blockFinalized(ctx, h)
+		}
+	}
+}
+
+// BlockFinalizedForTesting is used for testing	only.
+func (s *StopControl) BlockFinalizedForTesting(h *flow.Header) {
+	s.blockFinalized(irrecoverable.MockSignalerContext{}, h)
```

Comment:
I tried to do this, but I'm not sure I like it. 
This means that you have a testing "StopControl" (MockStopControl) that you use in your stop control tests, and that seems wrong, because you are testing the actual stop control, not a mock one. 

Also this would mean that engine_tests.go would have to pass around a "MockStopControl" which is also a bit weird.